File: app/src/main/java/nl/sjtek/client/android/cards/MusicSheetCard.java
package nl.sjtek.client.android.cards;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import nl.sjtek.client.android.R;
import nl.sjtek.client.android.api.API;
import nl.sjtek.client.android.api.Action;
import nl.sjtek.control.data.responses.MusicResponse;
import nl.sjtek.control.data.responses.ResponseCollection;

public class MusicSheetCard extends LinearLayout {

    @BindView(R.id.textViewTopTitle)
    TextView textViewTopTitle;
    @BindView(R.id.textViewTopArtist)
    TextView textViewTopArtist;
    @BindView(R.id.textViewTitle)
    TextView textViewTitle;
    @BindView(R.id.textViewArtist)
    TextView textViewArtist;
    @BindView(R.id.buttonPlay)
    ImageButton imageButtonPlay;
    @BindView(R.id.buttonTopPlay)
    ImageButton imageButtonTopPlay;
    @BindView(R.id.imageViewAlbumArt)
    ImageView imageViewAlbumArt;
    @BindView(R.id.imageViewArtistArt)
    ImageView imageViewArtistArt;
    @BindView(R.id.viewTitle)
    View viewTitle;
    @BindView(R.id.viewInfo)
    View viewInfo;

    private SheetClickListener sheetClickListener;

    public MusicSheetCard(Context context) {
        this(context, null);
    }

    public MusicSheetCard(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MusicSheetCard(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        inflate(context, R.layout.card_music_sheet, this);
        ButterKnife.bind(this);
        viewTitle.setAlpha(0);
    }

    public void setSheetClickListener(SheetClickListener sheetClickListener) {
        this.sheetClickListener = sheetClickListener;
    }


    public void onSlide(float slideOffset) {
        viewTitle.setAlpha(slideOffset);
        viewInfo.setAlpha(1 - slideOffset);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        EventBus.getDefault().register(this);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        EventBus.getDefault().unregister(this);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onUpdate(ResponseCollection responseCollection) {
        MusicResponse music = responseCollection.getMusic();
        setMusicInfo(music);
    }

    @OnClick({R.id.buttonPrevious})
    public void onPreviousClick() {
        API.action(getContext(), Action.Music.PREVIOUS);
    }

    @OnClick({R.id.buttonNext, R.id.buttonTopNext})
    public void onNextClick() {
        API.action(getContext(), Action.Music.NEXT);
    }

    @OnClick({R.id.buttonPlay, R.id.buttonTopPlay})
    public void onToggleClick() {
        API.action(getContext(), Action.Music.TOGGLE);
    }

    @OnClick(R.id.viewTop)
    public void onTopClick() {
        if (sheetClickListener != null) sheetClickListener.onSheetClick();
    }

    private void setMusicInfo(MusicResponse music) {
        MusicResponse.Song song = music.getSong();

        int buttonImage;
        if (music.getState() == MusicResponse.State.STATUS_PLAYING) {
            buttonImage = R.drawable.ic_pause_black_24dp;
        } else {
            buttonImage = R.drawable.ic_play_arrow_black_24dp;
        }

        imageButtonPlay.setImageResource(buttonImage);
        imageButtonTopPlay.setImageResource(buttonImage);

        textViewTopTitle.setText(song.getTitle());
        textViewTopArtist.setText(TextUtils.isEmpty(song.getArtist()) ? "Music stopped" : song.getArtist());

        textViewTitle.setText(song.getTitle());
        textViewArtist.setText(TextUtils.isEmpty(song.getArtist()) ? "Music stopped" : song.getArtist());

        String image = "";
        if (!TextUtils.isEmpty(song.getAlbum())) {
            image = song.getAlbumArt();
        } else if (!TextUtils.isEmpty(song.getArtistArt())) {
            image = song.getArtistArt();
        }

        if (TextUtils.isEmpty(image)) {
            imageViewAlbumArt.setImageDrawable(null);
            imageViewArtistArt.setImageDrawable(null);
        } else {
            Picasso.with(getContext())
                    .load(image)
                    .into(imageViewAlbumArt);

//            Picasso.with(getContext())
//                    .load(song.getArtistArt())
//                    .fit()
//                    .centerCrop()
//                    .transform(new BlurTransformation(getContext(), 20))
//                    .into(imageViewArtistArt);
        }
    }

    public interface SheetClickListener {
        void onSheetClick();
    }
}

File: app/src/main/java/nl/sjtek/client/android/cards/MusicSheetCard.kt
package nl.sjtek.client.android.cards

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.util.Pair
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.card_music_sheet.view.*
import nl.sjtek.client.android.R
import nl.sjtek.client.android.activities.ActivityMusic
import nl.sjtek.client.android.api.API
import nl.sjtek.control.data.actions.Actions
import nl.sjtek.control.data.parsers.ResponseHolder
import nl.sjtek.control.data.response.Music
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * Bottom sheet for controlling the music. Can be dragged up.
 */
class MusicSheetCard @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : LinearLayout(context, attrs, defStyleAttr) {

    var sheetClickListener: SheetClickListener? = null

    init {
        View.inflate(context, R.layout.card_music_sheet, this)
        viewTitle.alpha = 0f
        buttonPrevious.setOnClickListener(this::onPreviousClick)
        buttonNext.setOnClickListener(this::onNextClick)
        buttonTopNext.setOnClickListener(this::onNextClick)
        buttonPlay.setOnClickListener(this::onToggleClick)
        buttonTopPlay.setOnClickListener(this::onToggleClick)
        viewTop.setOnClickListener(this::onTopClick)
        buttonStartPlaylist.setOnClickListener(this::onStartPlaylistClick)
        imageViewAlbumArt.setOnClickListener(this::onStartPlaylistClick)
    }

    fun onSlide(slideOffset: Float) {
        viewTitle.alpha = slideOffset
        viewInfo.alpha = 1 - slideOffset
        val topButtonVisibility = if (slideOffset == 1f) View.GONE else View.VISIBLE
        buttonTopPlay.visibility = topButtonVisibility
        buttonTopNext.visibility = topButtonVisibility
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        EventBus.getDefault().register(this)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        EventBus.getDefault().unregister(this)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onUpdate(responseCollection: ResponseHolder) {
        val music = responseCollection.music
        setMusicInfo(music)
    }

    fun onPreviousClick(view: View? = null) = API.action(context, Actions.music.previous())
    fun onNextClick(view: View? = null) = API.action(context, Actions.music.next())
    fun onToggleClick(view: View? = null) = API.action(context, Actions.music.toggle())
    fun onStartPlaylistClick(view: View? = null) {
        val pair = arrayOf(
                Pair.create<View, String>(imageViewAlbumArt, "albumArt"),
                Pair.create<View, String>(textViewArtist, "trackArtist"),
                Pair.create<View, String>(textViewTitle, "trackTitle"),
                Pair.create<View, String>(buttonPlay, "buttonPlay"),
                Pair.create<View, String>(buttonPrevious, "buttonPrevious"),
                Pair.create<View, String>(buttonNext, "buttonNext"))
        val activity = context as Activity
        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(activity, *pair)
        context.startActivity(Intent(context, ActivityMusic::class.java), options.toBundle())
    }

    fun onTopClick(view: View? = null) {
        sheetClickListener?.onSheetClick()
    }

    private fun setMusicInfo(music: Music) {
        val buttonImage: Int = if (music.state == Music.State.PLAYING) {
            R.drawable.ic_pause_black_24dp
        } else {
            R.drawable.ic_play_arrow_black_24dp
        }

        buttonPlay.setImageResource(buttonImage)
        buttonTopPlay.setImageResource(buttonImage)

        textViewTopTitle.text = music.name
        textViewTopArtist.text = if (TextUtils.isEmpty(music.artist)) "Music stopped" else music.artist

        textViewTitle.text = music.name
        textViewArtist.text = if (TextUtils.isEmpty(music.artist)) "Music stopped" else music.artist

        val image = when {
            music.albumArt.isNotBlank() -> music.albumArt
            music.artistArt.isNotBlank() -> music.artistArt
            else -> ""
        }

        if (TextUtils.isEmpty(image)) {
            imageViewAlbumArt.setImageDrawable(null)
            imageViewArtistArt.setImageDrawable(null)
        } else {
            Picasso.with(context)
                    .load(image)
                    .into(imageViewAlbumArt)
        }
    }

    interface SheetClickListener {
        fun onSheetClick()
    }
}
